In [1]:
import json

In [2]:
# loading data
data_path = "./src/res_test.json"

data = json.load(open(data_path, 'r'))

In [3]:
def local_frequency(data: dict, field: str) -> int:
    counter = 0

    for key in data.keys():
        if len(data[key][field]) > 0 or data[key][field] == "true":
            counter += 1

    return counter

# Import analysis

In this section we want to analysis how many times there are local imports inside a local scope. What we want to find is:
* how many times there are local imports inside a local scope

In [4]:
pkgs_w_local_imports = local_frequency(data, "local_import") # count the number of packages with local imports

print(f"There are {pkgs_w_local_imports} packages with local imports over the total of {len(data.keys())} packages({pkgs_w_local_imports/len(data.keys())*100:.2f}%).")

There are 1314 packages with local imports over the total of 10000 packages(13.14%).


# Inner function analysis

We propose the same analysis that we done for the imports, but this time we want to find the number of times there are local functions defined inside a local scope.

In [5]:
pkgs_w_local_functions = local_frequency(data, "inner_function") # count the number of packages with local functions

print(f"There are {pkgs_w_local_functions} packages with local imports over the total of {len(data.keys())} packages({pkgs_w_local_functions/len(data.keys())*100:.2f}%).")

There are 1512 packages with local imports over the total of 10000 packages(15.12%).


# Shadowing analysis

We want to see if the shadowing occurs into packages

In [ ]:
pkgs_w_shadowing = local_frequency(data, "shadowing")

print(f"There are {pkgs_w_shadowing} packages with local imports over the total of {len(data.keys())} packages({pkgs_w_shadowing/len(data.keys())*100:.2f}%).")

# Monkey patching analysis

We want to see if in the wild the monkey patching is used. There are several way to do:
* presence and/or usage of **patch decorator**
* presence and/or usage of **context manager**
* usage of **atypical function** for with
* executing a **variables** swap inside a class

In [ ]:
pkgs_w_patch_decorator_import = local_frequency(data, "patch_decorator_import")

print(f"There are {pkgs_w_patch_decorator_import} packages with local imports over the total of {len(data.keys())} packages({pkgs_w_patch_decorator_import/len(data.keys())*100:.2f}%).")

In [ ]:
pkgs_w_patch_decorator_usage = local_frequency(data, "patch_decorator_usage")

print(f"There are {pkgs_w_patch_decorator_usage} packages with local imports over the total of {len(data.keys())} packages({pkgs_w_patch_decorator_usage/len(data.keys())*100:.2f}%).")

In [ ]:
pkgs_w_contextmanager_import = local_frequency(data, "contextmanager_import")

print(f"There are {pkgs_w_contextmanager_import} packages with local imports over the total of {len(data.keys())} packages({pkgs_w_contextmanager_import/len(data.keys())*100:.2f}%).")

In [ ]:
pkgs_w_contextmanager_usage = local_frequency(data, "contextmanager_usage")

print(f"There are {pkgs_w_contextmanager_usage} packages with local imports over the total of {len(data.keys())} packages({pkgs_w_contextmanager_usage/len(data.keys())*100:.2f}%).")

In [ ]:
pkgs_w_with_statement = local_frequency(data, "with_statement")

print(f"There are {pkgs_w_with_statement} packages with local imports over the total of {len(data.keys())} packages({pkgs_w_with_statement/len(data.keys())*100:.2f}%).")

In [ ]:
pkgs_w_overwrite_method_class = local_frequency(data, "overwrite_method_class")

print(f"There are {pkgs_w_overwrite_method_class} packages with local imports over the total of {len(data.keys())} packages({pkgs_w_overwrite_method_class/len(data.keys())*100:.2f}%).")

# Scope chain mean length

Now we want to investigate which is the average length of the scope chain over the most popular packages

In [1]:
mean = 0

for key in data.keys():
    mean += data[key]["scope_chain_length"]

mean /= len(data.keys())

print(f"Mean chain length: {mean}")

NameError: name 'data' is not defined

# Conclusions

In general, the practice of defining local imports and functions respectively in the 13.14% and 15.12% of the packages taken into analysis. Generally the average length of the scope chain is 2.8